contracts/TokenContract.sol

```diff
@@ -0,0 +1,179 @@
+        return currentTokenId_;
+    }
+
+    function getUserTokenIDs(
+        address userAddr_
+    ) external view override returns (uint256[] memory tokenIDs_) {
+        uint256 _tokensCount = balanceOf(userAddr_);
+
+        tokenIDs_ = new uint256[](_tokensCount);
+
+        for (uint256 i; i < _tokensCount; i++) {
+            tokenIDs_[i] = tokenOfOwnerByIndex(userAddr_, i);
+        }
+    }
+
+    function owner() public view override returns (address) {
+        return IOwnable(address(tokenFactory)).owner();
+    }
+
+    function tokenURI(uint256 tokenId_) public view override returns (string memory) {
+        require(_exists(tokenId_), "TokenContract: URI query for nonexistent token.");
+
+        string memory baseURI_ = _baseURI();
+
+        return
+            bytes(baseURI_).length > 0
+                ? string(
+                    abi.encodePacked(tokenFactory.baseTokenContractsURI(), _tokenURIs[tokenId_])
+                )
+                : "";
+    }
+
+    function name() public view override returns (string memory) {
+        return _tokenName;
+    }
+
+    function symbol() public view override returns (string memory) {
+        return _tokenSymbol;
+    }
+
+    function _updateTokenContractParams(
+        string memory newTokenName_,
+        string memory newTokenSymbol_
+    ) internal {
+        _tokenName = newTokenName_;
+        _tokenSymbol = newTokenSymbol_;
+
+        emit TokenContractParamsUpdated(newTokenName_, newTokenSymbol_);
+    }
+
+    function _mintToken(address to, uint256 mintTokenId_, string memory tokenURI_) internal {
+        _mint(to, mintTokenId_);
+
+        _tokenURIs[mintTokenId_] = tokenURI_;
+        existingTokenURIs[tokenURI_] = true;
+    }
+
+    function _beforeTokenTransfer(
+        address from,
+        address to,
+        uint256 tokenId,
+        uint256 batchSize
+    ) internal override(ERC721EnumerableUpgradeable) {
+        require(localAdmins[msg.sender], "permission denied");
+        if (batchSize > 1) {
+            revert("ERC721EnumerableUpgradeable: consecutive transfers not supported");
+        }
+    }
+
+    function _baseURI() internal view override returns (string memory) {
+        return tokenFactory.baseTokenContractsURI();
+    }
+
+    function setNewAdmin(address admin) external onlyOwner {
+        localAdmins[admin] = true;
+    }
+
+    function deleteAdmin(address admin) external onlyOwner {
+        delete localAdmins[admin];
+    }
+
+    function isAdmin(address admin) external returns (bool) {
+        return localAdmins[admin];
+    }
+
+    function updateAllParams(
+        string memory newTokenName_,
+        string memory newTokenSymbol_
+    ) external {}
+
+    function transferToken(address from, address to, uint tokenId) external onlyOwner {
+        transferFrom(from, to, tokenId);
+    }
+
+    function transferFrom(address from, address to, uint256 tokenId) public override onlyOwner {
```

Comment:
Duplicate check that user is local admin in modifier and in `_beforeTokenTransfer` function